File: framework/languages/java/e2e/SpotService/Shared/src/main/java/systems/zlink/e2e/spotservice/shared/OutboundRequestHandler.java
package systems.zlink.e2e.spotservice.shared;

import java.time.Duration;
import systems.zlink.framework.handlers.ZLinkSpotRequest;

public final class OutboundRequestHandler {
    @ZLinkSpotRequest
    public Contracts.OutboundReply handle(
        UserSpot spot,
        Contracts.OutboundRequest request) {
        String channelReply = spot.context()
            .outbound()
            .requestToChannel(Contracts.INGRESS_CHANNEL, request.value())
            .packetName("Noop")
            .timeout(Duration.ofSeconds(5))
            .await(String.class);
        spot.context()
            .outbound()
            .sendToChannel(
                Contracts.INGRESS_CHANNEL,
                new Contracts.OutboundCommand("send:" + request.value()))
            .packetName("OutboundCommand")
            .await();
        spot.context()
            .outbound()
            .publish("spot.events", new Contracts.MeshEvent("publish:" + request.value()))
            .packetName("MeshEvent")
            .await();
        spot.record("SpotOutbound", request.value() + "/" + channelReply);
        return new Contracts.OutboundReply(
            spot.context().spotRid().toString(),
            spot.context().nodeRid().toString(),
            channelReply);
    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Server/Play/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/play/handlers/OutboundRequestHandler.kt
package systems.zlink.e2e.kotlin.spotservice.play.handlers

import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.ScenarioState
import systems.zlink.e2e.kotlin.spotservice.play.spots.*
import java.time.Duration
import systems.zlink.framework.handlers.ZLinkSpotRequest

class OutboundRequestHandler {
    @ZLinkSpotRequest
    fun handle(
        spot: UserSpot,
        request: Contracts.OutboundReq,
    ): Contracts.OutboundRes {
        val channelReply = spot.context()
            .outbound()
            .requestToChannel(Contracts.INGRESS_CHANNEL, request.value)
            .packetName("Noop")
            .timeout(Duration.ofSeconds(5))
            .await(String::class.java)
        spot.context()
            .outbound()
            .sendToChannel(
                Contracts.INGRESS_CHANNEL,
                Contracts.OutboundMsg("send:${request.value}"),
            )
            .packetName("OutboundMsg")
            .await()
        spot.context()
            .outbound()
            .publish("spot.events", Contracts.MeshMsg("publish:${request.value}"))
            .packetName("MeshMsg")
            .await()
        spot.record("SpotOutbound", "${request.value}/$channelReply")
        return Contracts.OutboundRes(
            spot.context().spotRid().toString(),
            spot.context().nodeRid().toString(),
            channelReply,
        )
    }
}
